### PR TITLE
SANS2D: set DISP in top layer

### PR DESCRIPTION
**Pull request by RaiBishal**
In SANS2D `DISP` was being set in motor layer and vacuum tank OPI layer causing motor to ignore any `STOP` command when it is moving. I have removed the `DISP` in motor layer and now but it is still present in OPI layer to avoid Collisions.

### CODE REVIEW

**Comment by FreddieAkeroyd** [comment]
I notice calcout records like
```
field(INPA, "$(P)BEAMSTOP2Y:MTR.DMOV CP")
field(INPB, "$(P)BEAMSTOP2Y:MTR.SPMG CP")
```
As SPMG is CP if you try to change SPMG it will get reset back. Is this still desired, or does the `DISP` change fix all the SANS2D issues?

**Comment by RaiBishal** [comment]
> I notice calcout records like
> 
> ```
> field(INPA, "$(P)BEAMSTOP2Y:MTR.DMOV CP")
> field(INPB, "$(P)BEAMSTOP2Y:MTR.SPMG CP")
> ```
SPMG will only be set to pause once the movement is completed.

> As SPMG is CP if you try to change SPMG it will get reset back. Is this still desired, or does the `DISP` change fix all the SANS2D issues?

we misunderstood what was happening. SPMG was not reset at all. We set `DISP` when the motors start moving, since the `DISP` was set we were unable to stop it. 

Surprisingly, `STOP ALL` from the device screen worked because `STOP ALL` triggers another record which will set `.STOP` to 1 for all axes in the tank. However, if directly caput `.STOP` to 1 it will respect the `DISP`. 

In conclusion, it looks like when the `DISP` is set to the motor layer, direct caput to `.STOP` will fail but if we use a wrapper PV that wraps `.STOP` and process it, it will work fine.

**Comment by FreddieAkeroyd** [comment]
Ah yes, there is "When Non-zero" in the calcout.  Is the `(B == 2 && A == 1)` part of the calculation needed? According to the motor record docs  `If "SPMG" has the value "Move," the motor record will reset SPMG to "Pause" when a motion completes.` I was worried that if you press move to start a move then for an instant you will satisfy `(B == 2 && A == 1)` and it may cancel the move before it has started, but it looks like DMOV gets set to 0 as soon as you specify a new position and so when entering values by hand it is unlikely to happen. However i am concerned that a script might be able to trigger this race condition, so could you confirm whether  `(B == 2 && A == 1)` is really needed and remove if not.

**Comment by RaiBishal** [comment]
> Ah yes, there is "When Non-zero" in the calcout. Is the `(B == 2 && A == 1)` part of the calculation needed? According to the motor record docs `If "SPMG" has the value "Move," the motor record will reset SPMG to "Pause" when a motion completes.` I was worried that if you press move to start a move then for an instant you will satisfy `(B == 2 && A == 1)` and it may cancel the move before it has started, but it looks like DMOV gets set to 0 as soon as you specify a new position and so when entering values by hand it is unlikely to happen. However i am concerned that a script might be able to trigger this race condition, so could you confirm whether `(B == 2 && A == 1)` is really needed and remove if not.

I see what you mean. `(B == 2 && A == 1)` is there because if you do `MOVE ALL` then it will set `SPMG` to `Move`for all the axes in the tank. We want to make sure that `SPMG` revert back to `Pause` for axes whose `SP` was not changed otherwise it will forever stay with `SPMG` = `Move` so there is the chance of collision.  I think the better option would be to only set `SPMG` to `Move` only for the axes whose `SP` was changed but I can't think of a way to do it.

**Comment by RaiBishal** [comment]
Seems like `LVAL` contains the last value, maybe I can implement something cleaner with this.

**Comment by FreddieAkeroyd** [comment]
It may not be a problem, maybe you can check from a script if setting a value and triggering spmg move always works or does sometimes get caught with DMOV change being missed

**Comment by RaiBishal** [comment]
> It may not be a problem, maybe you can check from a script if setting a value and triggering spmg move always works or does sometimes get caught with DMOV change being missed

Just tested it, DMOV change seemed to be picked up on time. Not sure if this would be any different in Instrument machine.

**Comment by FreddieAkeroyd** [comment]
probably good enough - i guess they shouldn't generally be scripting just this motor anyway as it is part of collision avoidance and we don't do dynamic feedback at the moment?

**Comment by RaiBishal** [comment]
> probably good enough - i guess they shouldn't generally be scripting just this motor anyway as it is part of collision avoidance and we don't do dynamic feedback at the moment?

We do have some dynamic feedback i.e. tell how far apart the 4 axes are from their adjacent axes.

**Review by FreddieAkeroyd — APPROVED** [review]
